detox/src/artifacts/screenshot/SimulatorScreenshotPlugin.js

```diff
@@ -38,23 +38,26 @@ class SimulatorScreenshotPlugin extends ScreenshotArtifactPlugin {
   }
 
   async onBeforeUninstallApp(event) {
-    const snapshots = [
-      ...Object.values(this.snapshots.fromTest),
-      ...Object.values(this.snapshots.fromSession)
-    ];
+    await this.api.requestIdleCallback(async () => {
+      const snapshots = [
+        ...Object.values(this.snapshots.fromTest),
+        ...Object.values(this.snapshots.fromSession)
+      ];
+
+      await Promise.all(snapshots.map(s => s && s.relocate()));
+    });
 
-    await Promise.all(snapshots.map(s => s && s.relocate()));
     await super.onBeforeUninstallApp(event);
   }
 
-  async _onInvokeFailure({ params }) {
+  _onInvokeFailure({ params }) {
     const artifacts = {
       visibilityFailingScreenshots: params.visibilityFailingScreenshotsURL,
       visibilityFailingRects: params.visibilityFailingRectsURL,
     };
 
     for (const [key, artifactPath] of Object.entries(artifacts)) {
-      if (!artifactPath || await fs.isDirEmpty(artifactPath)) {
+      if (!artifactPath || fs.isDirEmptySync(artifactPath)) {
```

Comment:
Since the callback is SYNC, it is a bad idea to use async functions.

That omission caused a side-effect when artifacts of the current tests, due to prolonged asynchronous execution of `_onInvokeFailure` callback, got assigned to the next test context.

I had to switch to `fs.readdirSync` under the hood, to block the execution in a reliable way for this callback.